File: src/main/kotlin/com/github/bin/command/SimpleCommand.java
package com.github.bin.command;

import com.github.bin.service.RoomConfig;
import lombok.RequiredArgsConstructor;
import lombok.val;

/**
 * @author bin
 * @since 2023/08/22
 */
@RequiredArgsConstructor
public abstract class SimpleCommand implements Command {
    protected final String first;

    @Override
    public final boolean invoke(RoomConfig roomConfig, String id, String msg) {
        val split = test(msg);
        if (split == null) {
            return false;
        }
        return handler(roomConfig, id, split);
    }

    protected String test(String split) {
        if (split.startsWith(first)) {
            return split.substring(first.length());
        } else {
            return null;
        }
    }

    abstract boolean handler(RoomConfig roomConfig, String id, String msg);
}

File: src/main/kotlin/com/github/bin/command/SimpleCommand.kt
package com.github.bin.command

import com.github.bin.service.RoomConfig

/**
 * @author bin
 * @version 1.0.0
 * @since 2023/7/9
 */
abstract class SimpleCommand(
        private val first: String
) : Command {

    override fun invoke(roomConfig: RoomConfig, id: String, msg: String): Boolean {
        val split = test(msg) ?: return false
        return handler(roomConfig, id, split)
    }

    open fun test(split: String): String? {
        return if (split.startsWith(first)) split.substring(first.length) else null
    }

    abstract fun handler(roomConfig: RoomConfig, id: String, msg: String): Boolean
}
